File: backend/tools/serp_api.py
import httpx
import os

BRIGHT_DATA_API_KEY = os.getenv("BRIGHT_DATA_API_KEY", "")
SERP_BASE = "https://api.brightdata.com/serp/google"
TIMEOUT = 20.0


async def search_serp(query: str, num_results: int = 10, date_filter: str | None = None) -> list[dict]:
    """Search via Bright Data SERP API. Falls back to DuckDuckGo if zone not configured."""
    params: dict = {"q": query, "num": num_results, "brd_json": "1"}
    if date_filter == "past_week":
        params["tbs"] = "qdr:w"
    elif date_filter == "past_month":
        params["tbs"] = "qdr:m"

    if BRIGHT_DATA_API_KEY:
        try:
            async with httpx.AsyncClient(timeout=TIMEOUT) as client:
                resp = await client.get(
                    SERP_BASE,
                    params=params,
                    headers={"Authorization": f"Bearer {BRIGHT_DATA_API_KEY}"},
                )
                resp.raise_for_status()
                return _parse_serp_response(resp.json())
        except (httpx.HTTPStatusError, Exception):
            pass  # Fall through to DuckDuckGo fallback

    return await _search_duckduckgo(query, num_results)


async def _search_duckduckgo(query: str, num_results: int) -> list[dict]:
    """Fallback search using DuckDuckGo Instant Answer API."""
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT, follow_redirects=True) as client:
            resp = await client.get(
                "https://api.duckduckgo.com/",
                params={"q": query, "format": "json", "no_html": "1", "skip_disambig": "1"},
                headers={"User-Agent": "CompeteIQ/1.0 GTM Intelligence Agent"},
            )
            data = resp.json()
        results = []
        for topic in data.get("RelatedTopics", [])[:num_results]:
            if "Text" in topic and "FirstURL" in topic:
                results.append({
                    "title": topic.get("Text", "")[:80],
                    "url": topic.get("FirstURL", ""),
                    "snippet": topic.get("Text", ""),
                    "source_date": None,
                    "relevance_signal": "duckduckgo",
                })
        if data.get("AbstractText"):
            results.insert(0, {
                "title": data.get("Heading", query),
                "url": data.get("AbstractURL", ""),
                "snippet": data.get("AbstractText", ""),
                "source_date": None,
                "relevance_signal": "duckduckgo_abstract",
            })
        return results
    except Exception:
        return []


def _parse_serp_response(data: dict) -> list[dict]:
    results = []
    for item in data.get("organic", []):
        results.append({
            "title": item.get("title", ""),
            "url": item.get("link", item.get("url", "")),
            "snippet": item.get("description", item.get("snippet", "")),
            "source_date": item.get("date"),
            "relevance_signal": item.get("sitelinks_title"),
        })
    return results

File: frontend/lib/tools/serp-api.ts
const BRIGHT_DATA_API_KEY = process.env.BRIGHT_DATA_API_KEY ?? '';
const SERP_BASE = 'https://api.brightdata.com/serp/google';

export interface SerpResult {
  title: string; url: string; snippet: string;
  source_date: string | null; relevance_signal: string | null;
}

export async function searchSerp(query: string, numResults = 10, dateFilter?: string | null): Promise<SerpResult[]> {
  const params = new URLSearchParams({ q: query, num: String(numResults), brd_json: '1' });
  if (dateFilter === 'past_week') params.set('tbs', 'qdr:w');
  if (dateFilter === 'past_month') params.set('tbs', 'qdr:m');

  if (BRIGHT_DATA_API_KEY) {
    try {
      const resp = await fetch(`${SERP_BASE}?${params}`, {
        headers: { 'Authorization': `Bearer ${BRIGHT_DATA_API_KEY}` },
        signal: AbortSignal.timeout(20_000),
      });
      if (resp.ok) return parseSerpResponse(await resp.json());
    } catch {}
  }
  return searchDuckDuckGo(query, numResults);
}

async function searchDuckDuckGo(query: string, numResults: number): Promise<SerpResult[]> {
  try {
    const params = new URLSearchParams({ q: query, format: 'json', no_html: '1', skip_disambig: '1' });
    const resp = await fetch(`https://api.duckduckgo.com/?${params}`, {
      headers: { 'User-Agent': 'CompeteIQ/1.0 GTM Intelligence Agent' },
      signal: AbortSignal.timeout(20_000),
      redirect: 'follow',
    });
    const data = await resp.json();
    const results: SerpResult[] = [];
    for (const topic of ((data.RelatedTopics ?? []) as Record<string, string>[]).slice(0, numResults)) {
      if (topic.Text && topic.FirstURL) {
        results.push({ title: topic.Text.slice(0, 80), url: topic.FirstURL, snippet: topic.Text, source_date: null, relevance_signal: 'duckduckgo' });
      }
    }
    if (data.AbstractText) {
      results.unshift({ title: data.Heading ?? query, url: data.AbstractURL ?? '', snippet: data.AbstractText, source_date: null, relevance_signal: 'duckduckgo_abstract' });
    }
    return results;
  } catch { return []; }
}

function parseSerpResponse(data: Record<string, unknown>): SerpResult[] {
  return ((data.organic as Record<string, unknown>[]) ?? []).map((item) => ({
    title: String(item.title ?? ''),
    url: String(item.link ?? item.url ?? ''),
    snippet: String(item.description ?? item.snippet ?? ''),
    source_date: item.date ? String(item.date) : null,
    relevance_signal: item.sitelinks_title ? String(item.sitelinks_title) : null,
  }));
}
